libmamba/include/mamba/util/weakening_map.hpp

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 20123 QuantStack and Mamba Contributors
+//
+// Distributed under the terms of the BSD 3-Clause License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+
+#include <stdexcept>
+
+#include <fmt/format.h>
+
+namespace mamba::util
+{
+    /**
+     * A Map wrapper that can weaken a key to find more matches.
+     *
+     * The API of a standard map is unmodified, only methods ending with "weaken" look for
+     * multiple keys.
+     * This can be understood as an extreme generalization of defaults: when a key is not found,
+     * the behaviour is to look for another key.
+     * The behaviour for generating the sequence of weaken keys is controled by the Weakener.
+     */
+    template <typename Map, typename Weakener>
+    class weakening_map : private Map
+    {
+    public:
+
+        using Base = Map;
+        using typename Base::key_type;
+        using typename Base::mapped_type;
+        using typename Base::value_type;
+        using typename Base::size_type;
+        using typename Base::iterator;
+        using typename Base::const_iterator;
+        using weakener_type = Weakener;
+
+        using Base::Base;
+        template <typename... Args>
+        weakening_map(weakener_type weakener, Args&&... args);
+
+        using Base::begin;
+        using Base::end;
+        using Base::cbegin;
+        using Base::cend;
+
+        using Base::empty;
+        using Base::size;
+        using Base::max_size;
+
+        using Base::clear;
+        using Base::insert;
+        using Base::insert_or_assign;
+        using Base::emplace;
+        using Base::emplace_hint;
+        using Base::try_emplace;
+        using Base::erase;
+        using Base::swap;
+        using Base::extract;
+        using Base::merge;
+
+        using Base::reserve;
+
+        using Base::at;
+
+        [[nodiscard]] auto at_weaken(const key_type& key) -> mapped_type&;
+        [[nodiscard]] auto at_weaken(const key_type& key) const -> const mapped_type&;
```

Comment:
Should this replace the `at` API? Asked differently, does the user needs both methods? Is there a need to distinguish between strict finding and weaken finding?

Reply:
I'd say it's still relevant, for instance if users want to remove an entry, they would more likely want an exact match.